src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -139,25 +84,48 @@ void mf::XWaylandServer::execl_xwayland(XWaylandSpawner const& spawner, int wl_c
             xwayland_path.c_str(),
             dsp_str.c_str(),
             "-rootless",
-            "-wm", wm_fd_str.c_str(),
+            "-wm", x11_wm_server.c_str(),
             "-terminate",
         };
 
     for (auto const& fd : spawner.socket_fds())
     {
-        XWaylandSpawner::set_cloexec(fd, false);
+        mf::XWaylandSpawner::set_cloexec(fd, false);
         args.push_back("-listen");
         // strdup() may look like a leak, but execvp() will trash all resource management
         args.push_back(strdup(std::to_string(fd).c_str()));
     }
 
-    args.push_back(NULL);
+    args.push_back(nullptr);
 
     execvp(xwayland_path.c_str(), const_cast<char* const*>(args.data()));
 }
 
-namespace
+auto fork_xwayland_process(
+    mf::XWaylandSpawner const& spawner,
+    std::string const& xwayland_path) -> mf::XWaylandServer::XWaylandProcess
 {
+    auto const wayland_pipe = make_socket_pair();
+    auto const x11_wm_pipe = make_socket_pair();
+
+    mir::log_info("Starting XWayland");
+    pid_t const xwayland_pid = fork();
+
+    switch (xwayland_pid)
+    {
+    case -1:
+        BOOST_THROW_EXCEPTION(std::system_error(errno, std::generic_category(), "Failed to fork XWayland process"));
+
+    case 0:
+        exec_xwayland(spawner, xwayland_path, wayland_pipe.client, x11_wm_pipe.server);
+        fprintf(stderr, "Failed to start XWayland, should be unreachable");
```

Comment:
If you used `perror("Failed to exec XWayland")` here we'd get the error message from `execvpe` thrown in for free.

Reply:
(It might be better to make that call in `exec_xwayland` immediately after the exec call, though)